minetest/minetest
# [NOSQUASH] TouchScreenGUI refactor

**Goal of the PR**
This PR refactors TouchScreenGUI code to make it more safe and maintainable.

**How does the PR work?**
There are two parts of this PR (separated by two commits):
1. Fixes memory leaks and code style issues.
2. Simplifies calculations by using better data type.

**Does it resolve any reported issue?**
There are no reported issue on this, but I noticed that some members/properties/variables were not removed correctly after its parent object is destroyed. There are also some repeating parts inside the code, mainly for the virtual joystick.

**Does this relate to a goal in [the roadmap](https://github.com/minetest/minetest/blob/master/doc/direction.md)?**
Yes, this PR tries to refactor internal code.

**If not a bug fix, why is this PR needed? What usecases does it solve?**
From my tests, there are indications of memory leaks when using TouchScreenGUI code.

## To do

This PR is Ready for Review.

## How to test

To test the change statically,
1. Read the changes in the code base.
2. Verify that the old code and the new code are the same functionally.

To test the change practically,
1. Run Minetest on a touch-screen device.
2. Play in a world or on a server.
3. There should be no behaviour changes when using touch-screen GUI.
4. Back to Menu.
5. There should be no crashes.

To spesifically test memory leaks indication,
1. Run Minetest (without this PR) on a touch-screen device.
2. Play in a world or on a server.
3. Wait until its RAM usage has settled.
4. Note down its RAM usage.
5. Back to Menu.
6. Repeat step number 2–5 for a few times.
7. Exit to OS.
8. Run Minetest (with this PR) on a touch-screen device.
9. Repeat step number 2–7 on this new version.
10. Compare the results.

**Android: APK builds**
- [armeabi-v7a](https://drive.google.com/file/d/1yZ-mgzrtyXfMvRU3twbs9vlgrIP9hffp/view)
- [arm64-v8a](https://drive.google.com/file/d/1Z5L-VAKwItUcsRf2vSHaoPebEek3btR3/view)
- [x86](https://drive.google.com/file/d/1c-0jRF5gRBiTixCKPOqMdwo6LbNz1v6k/view)
- [x86_64](https://drive.google.com/file/d/1MbpAyU3tiKUuYiEZethHfDB3H4LdNpjO/view)

<details>
<summary>Older APK builds</summary>

- [armeabi-v7a](https://drive.google.com/file/d/1OFhpUZZe_oBtpvdfdc2pImb_XajdNRyP/view)
- [arm64-v8a](https://drive.google.com/file/d/1CkGR9gcu2DW2loBTio7o0cbsoTgqz2K_/view)
- [x86](https://drive.google.com/file/d/1Zb1rEUAfnKKpzZMwk_4i65PFecp6isz-/view)
- [x86_64](https://drive.google.com/file/d/1Dds8mAxdNZ_6aJVq6ebaodpbzfygDBH6/view)
</details>
